contracts/integrations/test/framework/actors/keeper.ts

```diff
@@ -75,6 +84,51 @@ export function KeeperMixin<TBase extends Constructor>(Base: TBase): TBase & Con
                 ),
             );
         }
+
+        private async *_validFinalizePool(): AsyncIterableIterator<AssertionResult | void> {
+            const { stakingPools } = this.actor.simulationEnvironment!;
+            const assertion = validFinalizePoolAssertion(this.actor.deployment, this.actor.simulationEnvironment!);
+            while (true) {
+                // Finalize a random pool, or do nothing if there are no pools in the simulation yet.
+                const poolId = Pseudorandom.sample(Object.keys(stakingPools));
+                if (poolId === undefined) {
+                    yield;
+                } else {
+                    yield assertion.executeAsync([poolId], { from: this.actor.address });
+                }
+            }
+        }
+
+        private async *_validEndEpoch(): AsyncIterableIterator<AssertionResult | void> {
+            const assertion = validEndEpochAssertion(this.actor.deployment, this.actor.simulationEnvironment!);
+            const { stakingWrapper } = this.actor.deployment.staking;
+            while (true) {
+                const { currentEpoch } = this.actor.simulationEnvironment!;
+                const aggregatedStats = AggregatedStats.fromArray(
+                    await stakingWrapper.aggregatedStatsByEpoch(currentEpoch.minus(1)).callAsync(),
+                );
+                if (aggregatedStats.numPoolsToFinalize.isGreaterThan(0)) {
+                    // Can't end the epoch if the previous epoch is not fully finalized.
+                    yield;
+                } else {
+                    await this._fastForwardToNextEpochAsync();
+                    yield assertion.executeAsync([], { from: this.actor.address });
+                }
+            }
+        }
+
+        private async _fastForwardToNextEpochAsync(): Promise<void> {
+            const { stakingWrapper } = this.actor.deployment.staking;
+
+            // increase timestamp of next block by how many seconds we need to
+            // get to the next epoch.
```

Comment:
Is this logic removed from the staking api wrapper? O/w maybe we could call it here #DRY?

Reply:
I'd like to eventually get rid of the staking api wrapper altogether, so I'm going to leave as is for now